src/languages/java.js

```diff
@@ -18,32 +18,25 @@ export default function(hljs) {
   var ANNOTATION = {
     className: 'meta',
     begin: '@' + JAVA_IDENT_RE,
-    contains:[
+    contains: [
       {
         begin: /\(/,
         end: /\)/,
         contains: ["self"] // allow nested () inside our annotation
       },
     ]
   }
-  // https://docs.oracle.com/javase/7/docs/technotes/guides/language/underscores-literals.html
-  var JAVA_NUMBER_RE = '\\b' +
-    '(' +
-      '0[bB]([01]+[01_]+[01]+|[01]+)' + // 0b...
-      '|' +
-      '0[xX]([a-fA-F0-9]+[a-fA-F0-9_]+[a-fA-F0-9]+|[a-fA-F0-9]+)' + // 0x...
-      '|' +
-      '(' +
-        '([\\d]+[\\d_]+[\\d]+|[\\d]+)(\\.([\\d]+[\\d_]+[\\d]+|[\\d]+))?' +
-        '|' +
-        '\\.([\\d]+[\\d_]+[\\d]+|[\\d]+)' +
-      ')' +
-      '([eE][-+]?\\d+)?' + // octal, decimal, float
-    ')' +
-    '[lLfF]?';
   var JAVA_NUMBER_MODE = {
     className: 'number',
-    begin: JAVA_NUMBER_RE,
+    variants: [
+      { begin: '\\b(0[bB][01]+[01_]*[01]+)[lL]?' }, // binary
```

Comment:
Please add tests for single digit variants of all these... I'm pretty sure you aren't handling that now.  For example:

`0b0` and `0b1` both fails with this regex.

Octal looks to have the same issue, I didn't look closely at the others.

Reply:
I think you just need to make the following part optional:

`\\b(0[bB][01]+([01_]*[01]+)?)[lL]?`

Or some such...